block_device.go

```diff
@@ -34,11 +35,59 @@ type BlockDeviceGetSectorSizeFunc func(file *os.File) (int, error)
 // BlockDevice is a low-level wrapper for a block device. The wrapper implements
 // io.Writer and io.Closer interfaces.
 type BlockDevice struct {
-	Path      string               // device path, ex. /dev/mmcblk0p1
-	out       *os.File             // os.File for writing
-	w         *utils.LimitedWriter // wrapper for `out` limited the number of bytes written
-	typeUBI   bool                 // Set to true if we are updating an UBI volume
-	ImageSize int64                // image size
+	Path               string               // device path, ex. /dev/mmcblk0p1
+	out                *os.File             // os.File for writing
+	w                  *utils.LimitedWriter // wrapper for `out` limited the number of bytes written
+	typeUBI            bool                 // Set to true if we are updating an UBI volume
+	ImageSize          int64                // image size
+	FlushIntervalBytes uint64               // Force a flush to disk each time this many bytes are written
+}
+
+// A WriteSyncer is an io.Writer that also implements a Sync() function which commits written data to stable storage.
+// For instance, an os.File is a WriteSyncer.
+type WriteSyncer interface {
+	io.Writer
+	Sync() error // Commits previously-written data to stable storage.
+}
+
+// FlushingWriter is a wrapper around a WriteSyncer which forces a Sync() to occur every so many bytes.
+// FlushingWriter implements WriteSyncer.
+type FlushingWriter struct {
+	WF                    WriteSyncer
+	FlushIntervalBytes    uint64
+	unflushedBytesWritten uint64
+}
+
+// NewFlushingWriter returns a WriteSyncer which wraps the provided WriteSyncer
+// and automatically flushes (calls Sync()) each time the specified number of
+// bytes is written.
+// Setting flushIntervalBytes == 0 causes Sync() to be called after every Write().
+func NewFlushingWriter(wf WriteSyncer, flushIntervalBytes uint64) *FlushingWriter {
+	return &FlushingWriter{
+		WF:                    wf,
+		FlushIntervalBytes:    flushIntervalBytes,
+		unflushedBytesWritten: 0,
+	}
+}
+
+func (fw *FlushingWriter) Write(p []byte) (int, error) {
+	rv, err := fw.WF.Write(p)
+
+	fw.unflushedBytesWritten += uint64(rv)
+
+	if err != nil {
+		return rv, err
+	} else if fw.unflushedBytesWritten >= fw.FlushIntervalBytes {
+		err = fw.Sync()
+	}
+
+	return rv, err
+}
+
+func (fw *FlushingWriter) Sync() error {
+	err := fw.WF.Sync()
+	fw.unflushedBytesWritten = 0
```

Comment:
Note: The current use of `io.CopyBuffer` in `device.go` does not guarantee that data is actually written exactly in chunks of a single sector (because of the convention that a Go `io.Reader`'s `Read()` returns what is "currently available").  In practice, by observing the `write` syscall via `strace`, it usually is a single sector (512 bytes in my case), but that behavior is not actually guaranteed by the current code, so it's currently already possible to get "out of sync" with the sector boundaries.

I also believe that calling `write` (the syscall) in increments of single sectors was contributing to the prevalence of the mmc OOM condition.  My new pull request ( #342 ) addresses this and guarantees that we always `write` in "medium size" increments.